### PR TITLE
[stable27] pruneOutdatedSyncTokens deletes all entries

### PR DESCRIPTION
**Pull request by tcitworld**
Backport of https://github.com/nextcloud/server/pull/38639/

### CODE REVIEW

**Review by st3iny — APPROVED** [review]

**Review by miaulalala — APPROVED** [review]

**Comment by st3iny** [comment]
Cypress failures are unrelated.

File: apps/dav/lib/CalDAV/CalDavBackend.php [hunk 1]
```diff
@@ -3134,10 +3134,19 @@ public function pruneOutdatedSyncTokens(int $keep = 10_000): int {
 		if ($keep < 0) {
 			throw new \InvalidArgumentException();
 		}
+
+		$query = $this->db->getQueryBuilder();
+		$query->select($query->func()->max('id'))
+			->from('calendarchanges');
+
+		$maxId =  $query->executeQuery()->fetchOne();
```
Comment (nickvergessen) [inline apps/dav/lib/CalDAV/CalDavBackend.php:3142]:
```suggestion
		$result = $query->executeQuery();
		$maxId = (int) $result->fetchOne();
		$result->closeCursor();
```

File: apps/dav/lib/CardDAV/CardDavBackend.php [hunk 2]
```diff
@@ -1399,10 +1399,19 @@ public function pruneOutdatedSyncTokens(int $keep = 10_000): int {
 		if ($keep < 0) {
 			throw new \InvalidArgumentException();
 		}
+
+		$query = $this->db->getQueryBuilder();
+		$query->select($query->func()->max('id'))
+			->from('addressbookchanges');
+
+		$maxId =  $query->executeQuery()->fetchOne();
```
Comment (nickvergessen) [inline apps/dav/lib/CardDAV/CardDavBackend.php:1407]:
```suggestion
		$result = $query->executeQuery();
		$maxId = (int) $result->fetchOne();
		$result->closeCursor();
```

**Review by nickvergessen — APPROVED** [review]

File: apps/dav/lib/CalDAV/CalDavBackend.php (hunk 1 above)
Comment (st3iny) [reply apps/dav/lib/CalDAV/CalDavBackend.php]: Done

File: apps/dav/lib/CardDAV/CardDavBackend.php (hunk 2 above)
Comment (st3iny) [reply apps/dav/lib/CardDAV/CardDavBackend.php]: Done

**Pull request merged**